urdf_parser_plugin/include/urdf_parser_plugin/parser.h

```diff
@@ -54,7 +56,14 @@ class URDFParser
   }
 
   /// \brief Load Model from string
-  virtual urdf::ModelInterfaceSharedPtr parse(const std::string &xml_string) = 0;
+  /// \return nullptr and write to stderr if the given string is invalid
+  virtual urdf::ModelInterfaceSharedPtr parse(const std::string & data) = 0;
+
+  /// \brief Indicate if data is meant to be parsed by this parser
+  /// \return confidence that the data is meant for this parser, where smaller
+  ///         values mean more confidence, 0 is absolute certainty, and
+  ///         std::numeric_limits<size_t>::max() is no confidence.
```

Comment:
My intuition would be that a score of 1.0 is absolute certainty and 0.0 is no confidence (treating this as a probability), but that would require floating point comparisons. Is there a reason to avoid floating point? 

I would have trouble thinking what a score like `1000` would mean in this context. To back out a probability from this score would be something like `1-(x/std::numeric_limits<size_t>::max())` and plugging in our hypothetical `1000` in for `x`?

Reply:
>  Is there a reason to avoid floating point?

No reason to avoid floating point.

> (treating this as a probability)

I don't think the score can be treated as a probability. One could take a bunch of robot descriptions, run them through the `might_handle()` function, and count the number of URDF or non-URDF files in a range of scores. Then one could say "documents that scored between 29 and 42 were URDF files 50% of the time".

> I would have trouble thinking what a score like `1000` would mean in this context.

I don't think `1000` would mean anything on its own; it only has meaning when compared against other scores. If someone scores 10 free throws in a minute, are they an NBA basketball player? I don't think that has an answer. Now say there are two people and one scores 10 free throws in a minute while the other scores 2 - which one is more likely to be an NBA player? The latter is what's happening here.

Reply:
> (with a caveat that the free throw test isn't the same for each plugin since each has it's own might_handle() implementation).

I think this is where the intuition for a probability comes in. What seems most likely is that plugins are either going to score a `0` for "yes" or `std::numeric_limits<size_t>::max()` for "no", at which point anything in between will be a best guess.

Reply:
> What seems most likely is that plugins are either going to score a 0 for "yes" or std::numeric_limits<size_t>::max() for "no", at which point anything in between will be a best guess.

Yeah, the scores less than `std::numeric_limits<size_t>::max()` are a bit arbitrary. Maybe it would be more intuitive if the documentation was "return the position in the string where you became sure this document was for you"?

Reply:
> I think this is where the intuition for a probability comes in. What seems most likely is that plugins are either going to score a `0` for "yes" or `std::numeric_limits<size_t>::max()` for "no", at which point anything in between will be a best guess.

Yeah, and not only that, the writer of a new plugin won't know whether they'll be scored "better" than the other plugins without reading all other available plugins.  In which case, this regresses towards the binary 0/1 .  I'm now rethinking the whole plan to return anything but a boolean :).  That being said, it isn't something I would block on, though a more constrained range and/or better documentation on what exactly to return would be helpful.

Reply:
69b3d49138dfbbe8335ddda528f6251caf400bcf defines the score range being the position in the document the parser became confident the format was for it, and 025f816fb30fb2f55179309f186251c834da999b makes the urdf plugin return `data.size()` instead of the max size_t value.

From offline conversation on this topic, 31014c02b89ba234e13b4ffd8f5d71076720d8aa makes `urdf::Model` try the urdf plugin anyways if no plugin things the file is for it.